cirq-google/cirq_google/engine/abstract_local_job.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A helper for jobs that have been created on the Quantum Engine."""
+import copy
+import datetime
+
+from typing import Dict, List, Optional, Tuple, TYPE_CHECKING
+
+import cirq
+from cirq_google.engine import calibration
+from cirq_google.engine.abstract_job import AbstractJob
+
+if TYPE_CHECKING:
+    from cirq_google.engine.abstract_local_program import AbstractLocalProgram
+    from cirq_google.engine.abstract_local_processor import AbstractLocalProcessor
+    from cirq_google.engine.abstract_local_engine import AbstractLocalEngine
+
+
+class AbstractLocalJob(AbstractJob):
+    """A job that handles labels and descriptions locally in-memory.
+
+        This class is designed to make writing custom AbstractJob objects
+        that function in-memory easier.  This class will handle basic functionality
+        expected to be common across all local implementations.
+
+        Implementors of this class should write the following functions:
+          - Status functions: execution_status, failure
+          - Action functions: cancel, delete
+          - Result functions: results, batched_results, calibration_results
+    `
+        Attributes:
+          processor_ids: A string list of processor ids that this job can be run on.
+          processor_id: If provided, the processor id that the job was run on.
+              If not provided, assumed to be the first element of processor_ids
+          parent_program: Program containing this job
+          repetitions: number of repetitions for each parameter set
+          sweeps: list of Sweeps that this job should iterate through.
+    """
+
+    def __init__(
+        self,
+        *,
+        job_id: str,
+        parent_program: 'AbstractLocalProgram',
+        repetitions: int,
+        sweeps: List[cirq.Sweep],
+        processor_id: str = '',
+    ):
+        self._id = job_id
+        self._processor_id = processor_id
+        self._parent_program = parent_program
+        self._repetitions = repetitions
+        self._sweeps = sweeps
+        self._create_time = datetime.datetime.now()
+        self._update_time = datetime.datetime.now()
```

Comment:
What qualifies as an update? Should the `set_*` functions refresh this value?

Reply:
Good point, done.